File: sdk/examples/basic.rs
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! In this example we will create an evm transaction
//!
//! Rename `.env.example` to `.env` first, then run the command:
//! ```sh
//! cargo run --release --all-features --example basic
//! ```

use std::str::FromStr;

use instant::Duration;
use iota_sdk::{
    client::{
        constants::SHIMMER_COIN_TYPE,
        secret::{stronghold::StrongholdSecretManager, SecretManage, SecretManager},
        Client,
    },
    crypto::keys::bip39::Mnemonic,
    packable::PackableExt,
    types::block::{
        address::Bech32Address,
        output::{
            feature::{MetadataFeature, SenderFeature},
            unlock_condition::AddressUnlockCondition,
            BasicOutputBuilder, Feature,
        },
        payload::transaction::TransactionId,
        BlockId,
    },
    wallet::{account::types::AccountAddress, Account, ClientOptions},
    Wallet,
};
use iota_sdk_evm::{
    ethereum_agent_id, Api, ContractIdentity, EvmAddress, RequestMetadata, Result, ACCOUNTS, TESTNET_CHAIN_ADDRESS,
};
use url::Url;

#[tokio::main]
async fn main() -> Result<()> {
    // This example uses secrets in environment variables for simplicity which should not be done in production.
    dotenvy::dotenv().ok();

    // Setup Stronghold secret_manager
    let secret_manager = StrongholdSecretManager::builder()
        .password(std::env::var("STRONGHOLD_PASSWORD").unwrap())
        .build(std::env::var("STRONGHOLD_SNAPSHOT_PATH").unwrap())
        .unwrap();

    // Only required the first time, can also be generated with `manager.generate_mnemonic()?`
    let mnemonic = Mnemonic::from(std::env::var("MNEMONIC").unwrap());
    println!("Using mnemonic: {:?}", mnemonic);

    // secret_manager.store_mnemonic(mnemonic).await.unwrap();

    let client_options = ClientOptions::new()
        .with_node(&std::env::var("NODE_URL").unwrap())
        .unwrap();

    // Create the wallet
    let wallet = Wallet::builder()
        .with_secret_manager(SecretManager::Stronghold(secret_manager))
        .with_storage_path(&std::env::var("WALLET_DB_PATH").unwrap())
        .with_client_options(client_options)
        .with_coin_type(SHIMMER_COIN_TYPE)
        .finish()
        .await?;

    // Get or create a new account
    let account = wallet.get_or_create_account("Alice").await?;
    let account_addrs = account.generate_ed25519_addresses(2, None).await?;

    let balance = account.sync(None).await?;
    let account_addr = &account_addrs[0];
    println!("Using addr: '{:?}'", account_addr.address());

    let evm_address = wallet
        .get_secret_manager()
        .read()
        .await
        .generate_evm_addresses(
            iota_sdk::client::api::GetAddressesOptions::default()
                .with_range(*account_addr.key_index()..*account_addr.key_index() + 1),
        )
        .await?;
    let bytes: [u8; 20] = prefix_hex::decode(&evm_address[0]).unwrap();
    let _evm_addr = EvmAddress::from(bytes);

    println!("Using evm address: {:?}", evm_address);

    let wasp_url = std::env::var("WASP_NODE").unwrap();
    let api = Api::new(Url::parse(wasp_url.as_str()).unwrap());

    println!("wasp node: '{:?}'", api.info().await?);

    if balance.base_coin().available() > 0 {
        // 225053825 glow -> 220.053826 SMR ( 4999999 gas fee + 0.01 fee on evm )

        println!("Available balance: '{:?}'", balance.base_coin().available() / 2);

        // 56171331 -> 56143231
        // = 28100 = 28000 + MIN_GAS_FEE

        let assets_pre = api.get_balance(TESTNET_CHAIN_ADDRESS, *account_addr.address()).await?;
        println!("EVM balance pre: '{:?}'", assets_pre);

        let to_send = 1000; //balance.base_coin().available() / 2;
        println!("Sending: '{:?}'", to_send);
        // let _ = send_to_evm(&account, to_send, account_addr, Some(&evm_addr)).await?;
        let _ = send_to_evm(&account, to_send, account_addr, None).await?;

        // Wasp node updates after at most 1 more milestone
        println!("await 1 milestone...");
        one_milestone(account.client()).await?;

        let assets_post = api.get_balance(TESTNET_CHAIN_ADDRESS, *account_addr.address()).await?;
        println!("EVM balance post: '{:?}'", assets_post);

        println!("------[ WITHDRAW ]---------");

        let _ = withdraw_from_evm(&account, assets_post.base_tokens, account_addr).await?;

        // Wasp node updates after at most 1 more milestone
        println!("await 1 milestone...");
        one_milestone(account.client()).await?;

        let assets_post = api.get_balance(TESTNET_CHAIN_ADDRESS, *account_addr.address()).await?;
        println!("EVM balance post withdraw: '{:?}'", assets_post);
    } else {
        println!("no available balance. top up at '{:?}'", account_addr.address());
        iota_sdk::client::request_funds_from_faucet(&std::env::var("FAUCET_URL").unwrap(), account_addr.address())
            .await
            .unwrap();
    }

    Ok(())
}

async fn withdraw_from_evm(account: &Account, amount: u64, from_addr: &AccountAddress) -> Result<BlockId> {
    let protocol_parameters = account.client().get_protocol_parameters().await?;
    let metadata = withdraw(amount);
    let outputs = [
        BasicOutputBuilder::new_with_minimum_storage_deposit(protocol_parameters.rent_structure().clone())
            .add_unlock_condition(AddressUnlockCondition::from(
                Bech32Address::from_str(TESTNET_CHAIN_ADDRESS)?.inner().clone(),
            ))
            .with_features([
                Feature::from(MetadataFeature::new(metadata.pack_to_vec())?),
                Feature::from(SenderFeature::new(from_addr.address().clone())),
            ])
            .finish()
            .unwrap()
            .into(),
    ];

    let transaction = account.send_outputs(outputs, None).await?;
    println!(
        "Transaction sent: {}/transaction/{}",
        std::env::var("EXPLORER_URL").unwrap(),
        transaction.transaction_id
    );

    wait(account, &transaction.transaction_id).await
}

async fn send_to_evm(
    account: &Account,
    amount: u64,
    from_addr: &AccountAddress,
    to_address: Option<&EvmAddress>,
) -> Result<BlockId> {
    let protocol_parameters = account.client().get_protocol_parameters().await?;
    let metadata = match to_address {
        Some(a) => deposit_to(amount, a),
        None => deposit(amount),
    };

    let outputs = [
        BasicOutputBuilder::new_with_minimum_storage_deposit(protocol_parameters.rent_structure().clone())
            .add_unlock_condition(AddressUnlockCondition::from(
                Bech32Address::from_str(TESTNET_CHAIN_ADDRESS)?.inner().clone(),
            ))
            .with_features([
                Feature::from(MetadataFeature::new(metadata.pack_to_vec())?),
                Feature::from(SenderFeature::new(from_addr.address().clone())),
            ])
            .finish()
            .unwrap()
            .into(),
    ];

    let transaction = account.send_outputs(outputs, None).await?;
    println!(
        "Transaction sent: {}/transaction/{}",
        std::env::var("EXPLORER_URL").unwrap(),
        transaction.transaction_id
    );

    wait(account, &transaction.transaction_id).await
}

async fn one_milestone(_client: &Client) -> Result<()> {
    let duration = Duration::from_secs(3);
    tokio::time::sleep(duration).await;
    Ok(())
}

async fn wait(account: &Account, tx: &TransactionId) -> Result<BlockId> {
    // Wait for transaction to get included
    let block_id = account.retry_transaction_until_included(tx, None, None).await?;

    println!(
        "Block included: {}/block/{}",
        std::env::var("EXPLORER_URL").unwrap(),
        block_id
    );
    Ok(block_id)
}

fn withdraw(amount: u64) -> RequestMetadata {
    let mut metadata = RequestMetadata::new(
        ContractIdentity::Null,
        ACCOUNTS.to_string(),
        "withdraw".to_string(),
        500,
    );
    metadata.allowance.set_base_tokens(amount);
    metadata
}

fn deposit(amount: u64) -> RequestMetadata {
    let mut metadata = RequestMetadata::new(
        ContractIdentity::Null,
        ACCOUNTS.to_string(),
        "deposit".to_string(),
        iota_sdk_evm::MIN_GAS_FEE,
    );
    metadata.allowance.set_base_tokens(amount - iota_sdk_evm::MIN_GAS_FEE);

    metadata
}

fn deposit_to(amount: u64, address: &EvmAddress) -> RequestMetadata {
    let mut metadata = RequestMetadata::new(
        ContractIdentity::Null,
        ACCOUNTS.to_string(),
        "transferAllowanceTo".to_string(),
        iota_sdk_evm::MIN_GAS_FEE,
    );
    metadata.params.insert(
        "a".to_string(),
        ethereum_agent_id(
            "42f7da9bdb55b3ec87e5ac1a1e6d88e16768663fde5eca3429eb6f579cc538ac",
            address,
        ),
    );
    metadata.allowance.set_base_tokens(amount - iota_sdk_evm::MIN_GAS_FEE);

    metadata
}

File: bindings/nodejs/examples/basic/basic.ts
/* eslint-disable @typescript-eslint/no-var-requires */

import {
    Api,
    Utils,
    initLogger,
    Constants,
    EvmAddress,
    NullIdentity,
    RequestMetadata,
    Contract,
} from '@iota/sdk-evm';

import {
    Wallet,
    Account,
    SecretManager,
    WalletOptions,
    CoinType,
    SenderFeature,
    Utils as SdkUtils,
    TransactionId,
    Client,
    initLogger as sdkInitLogger,
    AccountAddress,
} from '@iota/sdk';
import { AddressUnlockCondition, BlockId } from '@iota/sdk';
// Run with command:
// yarn run-example ./basic/basic.ts

//

// This example uses secrets in environment variables for simplicity which should not be done in production.
require('dotenv').config({ path: '.env' });

// TODO: Use UnitsHelper.MAGNITUDE_MAP["M"]
const ONE_MI = BigInt(1000000);

async function run(): Promise<void> {
    try {
        initLogger();
        sdkInitLogger();
        for (const envVar of [
            'MNEMONIC',
            'STRONGHOLD_SNAPSHOT_PATH',
            'STRONGHOLD_PASSWORD',
            'WALLET_DB_PATH',
            'NODE_URL',
            'WASP_NODE',
        ]) {
            if (!(envVar in process.env)) {
                throw new Error(
                    `.env ${envVar} is undefined, see .env.example`,
                );
            }
        }

        const strongholdSecretManager = {
            stronghold: {
                snapshotPath: process.env.STRONGHOLD_SNAPSHOT_PATH,
                password: process.env.STRONGHOLD_PASSWORD,
            },
        };
        const secretManager = new SecretManager(strongholdSecretManager);
        console.log('Using mnemonic:', process.env.MNEMONIC);

        // Store the mnemonic in the Stronghold snapshot, this needs to be done only the first time.
        // The mnemonic can't be retrieved from the Stronghold file, so make a backup in a secure place!
        //await secretManager.storeMnemonic(process.env.MNEMONIC!);

        const walletOptions: WalletOptions = {
            storagePath: process.env.WALLET_DB_PATH,
            clientOptions: {
                nodes: [process.env.NODE_URL!],
                ignoreNodeHealth: true,
            },
            coinType: CoinType.Shimmer,
            secretManager: strongholdSecretManager,
        };

        let wallet = new Wallet(walletOptions);

        let accounts = await wallet.getAccounts();
        let account;
        if (accounts.length == 0) {
            account = await wallet.createAccount({});
        } else {
            account = accounts[0];
        }

        let balance = await account.sync();

        //let accountAddrs = await account.generateEd25519Addresses(2);
        let accountAddrs = await account.addresses();
        let accountAddr = accountAddrs[0];
        console.log(`Using addr: '${accountAddr.address}'`);

        // Prefixed with 0x
        let evm_address = await secretManager.generateEvmAddresses({
            range: {
                start: accountAddr.keyIndex,
                end: accountAddr.keyIndex + 1,
            },
        });

        console.log(`Using evm address: '${evm_address}'`);

        let api = await Api.create(process.env.WASP_NODE as string);
        let client = await wallet.getClient();

        if (balance.baseCoin.available > ONE_MI) {
            console.log(`Available balance: '${balance.baseCoin.available}'`);

            let assetsPre = await api.getBalance(
                Constants.TESTNET_CHAIN_ADDRESS,
                accountAddr.address,
            );
            console.log('EVM balance pre:', assetsPre);

            console.log(`Sending: '${ONE_MI}'`);
            let _blockId = await sendToEVM(
                account,
                client,
                ONE_MI,
                accountAddr,
            );

            // Wasp node updates after at most 1 more milestone
            console.log('await 1 milestone...');
            await oneMilestone(await wallet.getClient());
            let balancePost = await account.sync();
            console.log(
                `Available balance POST: '${balancePost.baseCoin.available}'`,
            );
            console.log(
                `Available balance DIFF: '${
                    balance.baseCoin.available - balancePost.baseCoin.available
                }'`,
            );

            let assetsPost = await api.getBalance(
                Constants.TESTNET_CHAIN_ADDRESS,
                accountAddr.address,
            );
            console.log('EVM balance POST: ', assetsPost);
            console.log(
                'EVM balance DIFF: ',
                assetsPost.baseTokens - assetsPre.baseTokens,
            );

            console.log('------[ WITHDRAW ]---------');

            _blockId = await withdrawFromEvm(
                account,
                api,
                client,
                BigInt(assetsPre.baseTokens),
                accountAddr,
            );

            // Wasp node updates after at most 1 more milestone
            console.log('await 1 milestone...');
            await oneMilestone(await wallet.getClient());
            let balancePost2 = await account.sync();
            console.log(
                `Available balance POST2: '${balancePost2.baseCoin.available}'`,
            );
            console.log(
                `Available balance DIFF: '${
                    balancePost2.baseCoin.available - balance.baseCoin.available
                }'`,
            );

            let assetsPost2 = await api.getBalance(
                Constants.TESTNET_CHAIN_ADDRESS,
                accountAddr.address,
            );
            console.log('EVM balance post withdraw: ', assetsPost2);
            console.log(
                'EVM balance DIFF: ',
                assetsPost2.baseTokens - assetsPre.baseTokens,
            );
        } else {
            console.log('no available balance. top up at', accountAddr.address);
            client.requestFundsFromFaucet(
                process.env.FAUCET_URL!,
                accountAddr.address,
            );
        }
    } catch (err) {
        console.error(err);
    }
}

// Example translation for withdraw function
async function withdrawFromEvm(
    account: Account,
    api: Api,
    client: Client,
    amount: bigint,
    fromAddr: AccountAddress,
): Promise<BlockId> {
    const metadata = withdraw(amount);
    console.log(metadata);

    //let gasFee = await api.estimateGasOffLedger(Constants.TESTNET_CHAIN_ADDRESS, metadata);
    //console.log(gasFee);

    let outputs = [
        await client.buildBasicOutput({
            unlockConditions: [
                new AddressUnlockCondition(
                    SdkUtils.parseBech32Address(
                        Constants.TESTNET_CHAIN_ADDRESS,
                    ),
                ),
            ],
            features: [
                metadata.asFeature(),
                new SenderFeature(
                    SdkUtils.parseBech32Address(fromAddr.address),
                ),
            ],
        }),
    ];
    console.log(outputs);
    const minDeposit = await client.minimumRequiredStorageDeposit(outputs[0]);
    console.log(minDeposit);
    metadata.allowance.baseTokens += BigInt(minDeposit);
    console.log(metadata);
    console.log(metadata.asFeature());

    outputs = [
        await client.buildBasicOutput({
            amount: BigInt(minDeposit) + Constants.MIN_GAS_FEE, // Use gasFee instead
            unlockConditions: [
                new AddressUnlockCondition(
                    SdkUtils.parseBech32Address(
                        Constants.TESTNET_CHAIN_ADDRESS,
                    ),
                ),
            ],
            features: [
                metadata.asFeature(),
                new SenderFeature(
                    SdkUtils.parseBech32Address(fromAddr.address),
                ),
            ],
        }),
    ];
    console.log(outputs);

    const transaction = await account.sendOutputs(outputs);
    console.log(transaction);
    console.log(
        'Transaction sent:',
        `${process.env.EXPLORER_URL}/${transaction.transactionId}`,
    );

    return wait(account, transaction.transactionId);
}

// Example translation for withdraw function

function withdraw(amount: bigint): RequestMetadata {
    const metadata = new RequestMetadata(
        NullIdentity,
        Contract.Accounts,
        'withdraw',
        Constants.MIN_GAS_FEE * BigInt(100),
    );
    metadata.allowance.baseTokens = amount;
    return metadata;
}

async function sendToEVM(
    account: Account,
    client: Client,
    amount: bigint,
    fromAddr: AccountAddress,
    toAddress?: EvmAddress,
): Promise<BlockId> {
    const metadata = toAddress ? depositTo(amount, toAddress) : deposit(amount);

    let outputs = [
        await client.buildBasicOutput({
            amount: metadata.allowance.baseTokens,
            unlockConditions: [
                new AddressUnlockCondition(
                    SdkUtils.parseBech32Address(
                        Constants.TESTNET_CHAIN_ADDRESS,
                    ),
                ),
            ],
            features: [
                metadata.asFeature(),
                new SenderFeature(
                    SdkUtils.parseBech32Address(fromAddr.address),
                ),
            ],
        }),
    ];

    const transaction = await account.sendOutputs(outputs);
    console.log(
        `Transaction sent: ${process.env.EXPLORER_URL}/transaction/${transaction.transactionId}`,
    );

    return wait(account, transaction.transactionId);
}

async function oneMilestone(client: Client): Promise<void> {
    const duration = 3; // seconds
    await new Promise((resolve) => setTimeout(resolve, duration * 1000));
}

async function wait(account: Account, tx: TransactionId): Promise<BlockId> {
    // Wait for transaction to get included
    let blockId = await account.retryTransactionUntilIncluded(tx);

    console.log(`Block included: ${process.env.EXPLORER_URL}/block/${blockId}`);
    return blockId;
}

function deposit(amount: bigint): RequestMetadata {
    const metadata = new RequestMetadata(
        NullIdentity,
        Contract.Accounts,
        'deposit',
        Constants.MIN_GAS_FEE * BigInt(100),
    );
    metadata.allowance.baseTokens = amount;

    return metadata;
}

function depositTo(amount: bigint, address: EvmAddress): RequestMetadata {
    const metadata = new RequestMetadata(
        NullIdentity,
        Contract.Accounts,
        'transferAllowanceTo',
        Constants.MIN_GAS_FEE * BigInt(100),
    );
    metadata.params.set(
        'a',
        Utils.ethereumAgentId(
            '42f7da9bdb55b3ec87e5ac1a1e6d88e16768663fde5eca3429eb6f579cc538ac',
            address,
        ),
    );
    metadata.allowance.baseTokens = amount;

    return metadata;
}

void run().then(() => process.exit());
